tools/dockerfile/push_testing_images.sh

```diff
@@ -57,12 +69,70 @@ do
     DOCKER_IMAGE_TAG=$(sha1sum $DOCKERFILE_DIR/Dockerfile | cut -f1 -d\ )
   fi
 
-  # skip the image if it already exists in the repo 
-  curl --silent -f -lSL https://registry.hub.docker.com/v2/repositories/${DOCKERHUB_ORGANIZATION}/${DOCKER_IMAGE_NAME}/tags/${DOCKER_IMAGE_TAG} > /dev/null \
-      && continue
+  echo "Visiting ${DOCKERFILE_DIR}"
+
+  # SKIP_REMOTE controls whether artifact registry are going to be queried at all.
+  if [ "${SKIP_REMOTE}" == "" ]
+  then
+    DOCKER_IMAGE_DIGEST_REMOTE=$(gcloud artifacts docker images describe "${ARTIFACT_REGISTRY_PREFIX}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}" --format=json | jq -r '.image_summary.digest')
+
+    if [ "${DOCKER_IMAGE_DIGEST_REMOTE}" != "" ]
+    then
+      # skip building the image if it already exists in the destination registry
+      echo "Docker image ${DOCKER_IMAGE_NAME} already exists in artifact registry at the right version (tag ${DOCKER_IMAGE_TAG})."
+
+      # TODO: (sanity check) if remote check requested, check that the remote digest matches what's currently in the .current_version file.
+
+      # update info on what we consider to be the current version of the docker image (which will be used to run tests)
+      # we consider the sha256 image digest info from the artifact registry to be the canonical one
+      echo -n "${ARTIFACT_REGISTRY_PREFIX}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}@${DOCKER_IMAGE_DIGEST_REMOTE}" >${DOCKERFILE_DIR}.current_version
 
-  docker build -t ${DOCKERHUB_ORGANIZATION}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG} ${DOCKERFILE_DIR}
+      continue
+    fi
 
-  # "docker login" needs to be run in advance
-  docker push ${DOCKERHUB_ORGANIZATION}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}
+    # TODO: (sanity check) if remote check requested, fail here since not all images have been uploaded to artifact registry
+  fi
+
+  # if the .current_version file doesn't exist or it doesn't contain the right SHA checksum,
+  # it is out of date and we will need to rebuild the docker image locally.
+  LOCAL_BUILD_REQUIRED=""
+  grep "^${ARTIFACT_REGISTRY_PREFIX}/${DOCKER_IMAGE_NAME}:${DOCKER_IMAGE_TAG}@sha256:.*" ${DOCKERFILE_DIR}.current_version >/dev/null && LOCAL_BUILD_REQUIRED=true
```

Comment:
Shouldn't it be the opposite condition? If `.current_version` contains the right Dockerfile checksum, then LOCAL_BUILD_REQUIRED should be false. If you're trying to leverage short-circuit evaluation, I think you'd want an `||` here instead of an `&&`.